SimTKmath/Geometry/src/ContactGeometry.cpp

```diff
@@ -755,10 +770,169 @@ trackSeparationFromLine(const Vec3& pointOnLine,
+        const Vec3 g  = -calcSurfaceGradient(pointOnLine);
+        const Mat33 H = -calcSurfaceHessian(pointOnLine);
+
+        // Add a weight to the newton step to avoid large steps.
+        constexpr Real w = 0.5;
+
+        // Update alpha.
+        const Real step = dot(g, d) / (dot(d, H * d) + w);
+
+        // Stop when converged.
+        if (std::abs(step) < tolerance) {
+            break;
+        }
+
+        // Clamp the stepsize.
+        constexpr Real maxStep = 0.25;
+        alpha -= std::min(std::max(-maxStep, step), maxStep);
+
+        // Stop when leaving bounds.
+        if (alpha < 0. || alpha > 1.) {
+            break;
+        }
+    }
+
+    // Write the point on line nearest the surface.
+    alpha = std::max(0., std::min(alpha, 1.));
+    nearestPointOnLine = pointA + d * alpha;
+
+    const bool touchdown = -calcSurfaceValue(nearestPointOnLine) < tolerance;
+
+    SimTK_ASSERT_ALWAYS(iter < maxIter,
+        "Failed to compute point on line nearest "
+                                 "surface: Reached max iterations");
+
+    return touchdown;
+}
+
 //==============================================================================
 //                  GEODESIC EVALUATORS in CONTACT GEOMETRY
 //==============================================================================
 
+void ContactGeometry::shootGeodesicInDirectionImplicitly(
+    Vec3 pointApprox,
+    Vec3 tangentApprox,
+    Real finalArcLength,
+    Real& initStepSize,
+    Real integratorAccuracy,
+    Real constraintTolerance,
+    int maxIter,
+    Vec2& finalJacobi,
+    Vec2& finalJacobiDot,
+    std::function<void(const Real& l, const Vec3& x, const Vec3& t)>& log) const
+{
+    getImpl().shootGeodesicInDirectionImplicitly(pointApprox,
+    tangentApprox,
+    finalArcLength,
+    initStepSize,
+    integratorAccuracy,
+    constraintTolerance,
+    maxIter,
+    finalJacobi,
+    finalJacobiDot,
+    log);
+}
+
+void ContactGeometryImpl::shootGeodesicInDirectionImplicitly(
+    Vec3 pointApprox,
+    Vec3 tangentApprox,
+    Real finalArcLength,
+    Real& initStepSize,
+    Real integratorAccuracy,
+    Real constraintTolerance,
+    int maxIter,
+    Vec2& finalJacobi,
+    Vec2& finalJacobiDot,
+    std::function<void(const Real& l, const Vec3& x, const Vec3& t)>& log) const
+{
+    // integrator settings
+    const Real startArcLength = 0;
+
+    using Eqns = GeodesicOnImplicitSurface;
+
+    Eqns eqns(*this);
+    GeodesicIntegrator<GeodesicOnImplicitSurface>
+        integ(eqns,integratorAccuracy,constraintTolerance);
+    static const int N = GeodesicOnImplicitSurface::N;
+
+    integ.initialize(startArcLength, Eqns::getInitialState(pointApprox, UnitVec3(tangentApprox)));
+
+    // Aliases for the integrators internal time and state variables.
+    const Vec<N>& y = integ.getY();
+    const Real&   s = integ.getTime();  // arc length
+
+    // Simulate it, and record geodesic knot points after each step
+    int stepcnt = 0;
```

Comment:
see https://github.com/simbody/simbody/pull/791#discussion_r1663898202

Reply:
Same criticism applies. The design of someone else's API shouldn't prevent you from having a `constexpr int c_MaxIterations = 1000000` or similar as a failsafe to prevent threadlocks

Reply:
It doesnt prevent me, feel free to give feedback. I was only suggesting to refactor the GeodesicIntegrator API in another PR.